utils/pattern/src/single.rs

```diff
@@ -0,0 +1,256 @@
+///     Pattern::<SinglePlaceholder, _>::try_from_str("{0} days ago")
+///         .unwrap()
+///         .interpolate_to_string([5]),
+///     "5 days ago",
+/// );
+///
+/// // Single named placeholder:
+/// assert_eq!(
+///     Pattern::<SinglePlaceholder, _>::try_from_str("{name}")
+///         .unwrap()
+///         .interpolate_to_string(["Alice"]),
+///     "Alice",
+/// );
+///
+/// // No placeholder (note, the placeholder value is never accessed):
+/// assert_eq!(
+///     Pattern::<SinglePlaceholder, _>::try_from_str("yesterday")
+///         .unwrap()
+///         .interpolate_to_string(["hi"]),
+///     "yesterday",
+/// );
+///
+/// // Escaped placeholder and a real placeholder:
+/// assert_eq!(
+///     Pattern::<SinglePlaceholder, _>::try_from_str("'{0}' {1}")
+///         .unwrap()
+///         .interpolate_to_string(("hi",)),
+///     "{0} hi",
+/// );
+/// ```
+///
+/// [`Pattern::interpolate()`]: crate::Pattern::interpolate
+#[derive(Debug)]
+pub struct SinglePlaceholder {
+    _not_constructible: core::convert::Infallible,
+}
+
+impl crate::private::Sealed for SinglePlaceholder {}
+
+impl PatternBackend for SinglePlaceholder {
+    type PlaceholderKey = SinglePlaceholderKey;
+    type Store = str;
+    type Iter<'a> = SinglePlaceholderPatternIterator<'a>;
+
+    fn validate_store(store: &Self::Store) -> Result<(), Error> {
+        let placeholder_offset_char = store.chars().next().ok_or(Error::InvalidPattern)?;
+        let initial_offset = char::len_utf8(placeholder_offset_char);
+        let placeholder_offset = placeholder_offset_char as usize;
+        if placeholder_offset > store.len() - initial_offset + 1 {
+            return Err(Error::InvalidPattern);
+        }
+        Ok(())
+    }
+
+    fn iter_items(store: &Self::Store) -> Self::Iter<'_> {
+        let placeholder_offset_char = match store.chars().next() {
+            Some(i) => i,
+            None => {
+                debug_assert!(false);
+                '\0'
+            }
+        };
+        let initial_offset = char::len_utf8(placeholder_offset_char);
+        SinglePlaceholderPatternIterator {
+            store,
+            placeholder_offset: placeholder_offset_char as usize + initial_offset - 1,
+            current_offset: initial_offset,
+        }
+    }
+
+    #[cfg(feature = "alloc")]
+    fn try_from_items<
+        'a,
+        I: Iterator<Item = Result<PatternItemCow<'a, Self::PlaceholderKey>, Error>>,
+    >(
+        items: I,
+    ) -> Result<String, Error> {
+        let mut result = String::new();
+        let mut seen_placeholder = false;
+        for item in items {
+            match item? {
+                PatternItemCow::Literal(s) => result.push_str(&s),
+                PatternItemCow::Placeholder(_) if !seen_placeholder => {
+                    seen_placeholder = true;
+                    let placeholder_offset =
+                        u32::try_from(result.len() + 1).map_err(|_| Error::InvalidPattern)?;
```

Comment:
nit: ideally a placeholder at index 56000 should not error, you'd have to shift the value up into the encodable range

Reply:
Well ideally we wouldn't randomly reject pattern lengths in the surrogate range while accepting the others. I don't want to complicate the code by doing arithmetic to bring the varint into range. I would be okay if `str` allowed Generalized UTF-8 but it doesn't. If we really need this in the future I would rather make a pattern backend with `[u8]` that we enforce as Generalized UTF-8.

Reply:
(In this PR, I changed it to just reject placeholder offsets that are too big)